[WIP] In Memory Editor

333fred opened the pull request:
This is a fairly large change, so I want to get the review started now. The salient points are:
* Project file editing now uses an in-memory buffer instead of a temporary file. This greatly improves the experience, and `Open Containing Folder` now works as expected.
* Previous to this change, the state was all contained in a couple of files, and did not have defined state transitions. With this new change, that quickly became untenable. I've factored the state machine into a single class, EditorStateModel, which coordinates all state changes. I've tried to make the states descriptive, and provide examples and valid state transitions.

Main todos:
* I'm not handling saving dirty state at all when closing the buffer or updating the project file.
* Tests probably don't compile, let alone run.

I'm not expecting this PR to pass tests yet (or even for the tests to compile), those are still to come. Tagging @dotnet/project-system, @davkean, @BillHiebert, @srivatsn to start the review.

davkean commented on src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Editor/EditorStateModel.cs:
```diff
@@ -0,0 +1,302 @@
+            {
+                _currentState = _currentState == EditorState.EditorDirty ?
+                    EditorState.ProjectFileUpdateScheduledFromDirty : EditorState.ProjectFileUpdateScheduledFromClean;
+                _threadingService.JoinableTaskFactory.RunAsync(UpdateProjectFileAsync);
```
Where are exceptions from this observed? What's the default behavior of JTF on unobserved exceptions?

davkean commented on src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Editor/EditorStateModel.cs:
```diff
@@ -0,0 +1,302 @@
+        private async Task SetBufferReadonlyAsync(bool readOnly)
+        {
+            await _threadingService.SwitchToUIThread();
+            _textBufferManager.SetStateFlags(readOnly ? (uint)BUFFERSTATEFLAGS.BSF_USER_READONLY : 0);
```
Should we be reading the "current" state flags here and setting them, otherwise we're overwriting them here?

davkean commented on src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Editor/Listeners/TextBufferStateListener.cs:
```diff
@@ -0,0 +1,146 @@
+
+                // If the buffer isn't dirty, then we force the gutter green. The gutter goes green when the FileActionOccurred event is triggered.
+                // In the future, the editor will be able to expose an API for us to hook here, but until then, we manually reflect in and call
+                // the event.
```
Do we have a bug on this?

basoundr commented on src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Editor/EditorStateModel.cs:
```diff
@@ -0,0 +1,302 @@
+        public async Task OpenEditorAsync()
+        {
+            await _threadingService.SwitchToUIThread();
+            lock (_lock)
```
You are on the UIThread already. You dont need the locks

davkean replied:
I think the state can change off the UI thread.

davkean replied:
Yeah, see: CloseWindowAsync

davkean commented on src/Microsoft.VisualStudio.ProjectSystem.CSharp.VS/Packaging/CSharpProjectSystemPackage.cs:
```diff
@@ -20,6 +21,8 @@ namespace Microsoft.VisualStudio.Packaging
     [Guid(PackageGuid)]
     [PackageRegistration(AllowsBackgroundLoading = true, RegisterUsing = RegistrationMethod.CodeBase, UseManagedResourcesOnly = true)]
     [ProvideProjectFactory(typeof(MigrateXprojProjectFactory), null, "#8", "xproj", "xproj", null)]
+    [ProvideEditorFactory(typeof(LoadedProjectFileEditorFactory), 7, TrustLevel = __VSEDITORTRUSTLEVEL.ETL_HasUntrustedLogicalViews)]
```
Can we use named parameters? Magic numbers/strings are hard to read.

davkean commented on src/Microsoft.VisualStudio.ProjectSystem.CSharp.VS/Packaging/CSharpProjectSystemPackage.cs:
```diff
@@ -54,6 +57,7 @@ protected override Task InitializeAsync(CancellationToken cancellationToken, IPr
             _factory = new MigrateXprojProjectFactory(new ProcessRunner(), new FileSystem());
             _factory.SetSite(this);
             RegisterProjectFactory(_factory);
+            RegisterEditorFactory(new LoadedProjectFileEditorFactory(this));
```
Are you allowed to this on a background thread?

333fred replied:
It hasn't crashed anything yet, but probably not. I'll run a switch to the ui thread.

davkean commented on src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Editor/EditorStateModel.cs:
```diff
@@ -0,0 +1,302 @@
+            {
+                _windowFrame.Show();
+                _currentState = EditorState.EditorClean;
+                _projectFileModelWatcher.Initialize();
```
You initalize twice, why?

333fred replied:
A mistake. I had started to factor this part out so I could have the editorfactory actually start the initialization step and ran into the IVsWindowPane issue. I must have missed this when I put it back.

davkean commented on src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Editor/EditorStateModel.cs:
```diff
@@ -0,0 +1,302 @@
+
+            _projectFileModelWatcher?.Dispose();
+            _textBufferStateListener?.Dispose();
+            await (_frameEventsListener?.DisposeAsync() ?? Task.CompletedTask).ConfigureAwait(false);
```
I think a null check would be simpler to parse this in my mind.

davkean commented on src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Editor/EditorStateModel.cs:
```diff
@@ -0,0 +1,302 @@
+            /// * <see cref="ProjectFileUpdateScheduledFromClean"/> if the user makes a change from the UI, ushc as adding a NuGet package.
+            /// * <see cref="EditorClosing"/> if the user closes the buffer.
+            /// </summary>
+            EditorClean,
```
This state name needs to be more obvious, like "NoChanges" "NotDirty"

davkean commented on src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Editor/EditorStateModel.cs:
```diff
@@ -0,0 +1,302 @@
+                    _windowFrame?.Show();
+                    return;
+                }
+                _currentState = EditorState.Initializing;
```
I'd consider pull out the state changes into a separate method that validates and enforces the valid state changes as called out in the above docs.

davkean commented on src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Editor/EditorStateModel.cs:
```diff
@@ -0,0 +1,302 @@
+                if (_currentState == EditorState.NoEditor) return;
+
+                // If there was no changes, then we just close the editor
+                if (_currentState == EditorState.EditorClean || _currentState == EditorState.ProjectFileChangingFromClean)
```
What's the relevance of the ProjectFileChangingFromClean state when compared to EditorClean? You don't seem to do anything different with this state.

333fred replied:
While in this state, changes to the editor are disallowed. It is effectively a no-op, as it's only used inside this method, but if something comes in from another thread or we add logic later we should know when we're actually running a project file update.

davkean commented on src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Editor/EditorStateModel.cs:
```diff
@@ -0,0 +1,302 @@
+                // If the current state isn't clean or dirty, some other change is happening, and we should not touch the current state
+                if (_currentState != EditorState.EditorClean && _currentState != EditorState.EditorDirty) return;
+
+                _currentState = _currentState == EditorState.EditorClean ? EditorState.EditorDirty : EditorState.EditorClean;
```
This logic looks wrong, are you missing  check to isDirty?

333fred replied:
Yeah, don't know what I was thinking.

davkean commented on src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Editor/VsTextBufferProviderAggregator.cs:
```diff
@@ -0,0 +1,59 @@
+namespace Microsoft.VisualStudio.ProjectSystem.VS.Editor
+{
+    [Export(ExportContractNames.VsTypes.ProjectNodeComExtension)]
+    [AppliesTo(ProjectCapability.CSharpOrVisualBasic)]
```
Isn't there a capability for this?

davkean replied:
OpenProjectFile

davkean commented on src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Utilities/ExportFactory/ProjectFileModelWatcherFactory.cs:
```diff
@@ -0,0 +1,26 @@
+namespace Microsoft.VisualStudio.ProjectSystem.VS.Utilities.ExportFactory
+{
+    [Export(typeof(IExportFactory<IProjectFileModelWatcher>))]
+    internal class ProjectFileModelWatcherFactory : IExportFactory<IProjectFileModelWatcher>
```
Why does this exist? Why not just use ExportFactory directly?

333fred replied:
Mocking for tests.

333fred replied:
Actually, I think I can make a generic implementation by having a series of static methods that return an IExportFactory of the correct type for each type that we need:

```C#
    internal class GenericExportFactory<T> : IExportFactory<T>
    {
        [Export(typeof(IExportFactory<EditProjectFileVsFrameEvents>))]
        public static IExportFactory<EditProjectFileVsFrameEvents> EditProjectFileVsFrameEventsExportFactory(
            ExportFactory<EditProjectFileVsFrameEvents> factory)
        {
            return new GenericExportFactory<EditProjectFileVsFrameEvents>(factory);
        }

        private readonly ExportFactory<T> _factory;

        private GenericExportFactory(ExportFactory<T> factory)
        {
            _factory = factory;
        }

        public T CreateExport()
        {
            throw new NotImplementedException();
        }
    }
```
Which would be preferable? Or do you have another solution for mocking ExportFactory?

davkean replied:
ExportFactory takes a func, you don't need to mock it.

davkean commented on src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Utilities/ExportFactory/TextBufferStateListenerExportFactory.cs:
```diff
@@ -0,0 +1,21 @@
+namespace Microsoft.VisualStudio.ProjectSystem.VS.Utilities.ExportFactory
+{
+    [Export(typeof(IExportFactory<ITextBufferStateListener>))]
+    internal class TextBufferStateListenerExportFactory : IExportFactory<ITextBufferStateListener>
```
Ditto, can you just use ExportFactory directly?

davkean commented on src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Utilities/ExportFactory/TextBufferStateListenerExportFactory.cs:
```diff
@@ -0,0 +1,21 @@
+using Microsoft.VisualStudio.ProjectSystem.VS.Editor.Listeners;
+using System.ComponentModel.Composition;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.Utilities.ExportFactory
```
Really, really hate this namespace - if you're going to keep these around, just put them next to the objects they are creating.

333fred commented on src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Editor/EditorStateModel.cs:
```diff
@@ -0,0 +1,302 @@
+            }
+        }
+
+        private async Task SetBufferReadonlyAsync(bool readOnly)
```
I pulled this out of the class entirely, it belonged in the TextBufferManager.

333fred commented on src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Editor/EditorStateModel.cs:
```diff
@@ -0,0 +1,302 @@
+            /// A project file update has been scheduled. Before the update began, the state was <see cref="EditorClean"/>.
+            /// Allowed Transitions: <see cref="ProjectFileChangingFromClean"/> when the editor update is run.
+            /// </summary>
+            ProjectFileUpdateScheduledFromClean,
```
I'm going to rename these to BufferUpdate...

333fred commented:
@davkean, I've addressed a number of your concerns. Not all, but the stuff around state handling. I also fixed a couple of bugs that were causing VS to crash.

333fred commented:
Work on InMemory will continued to be tracking on 333fred/roslyn-project-system#InMemoryEditor, but I'm going to close this for now. I'll open a new pull when it's closer to ready.
